[ConstraintSystem] Introduce a new type to represent a type hole

xedin opened the pull request:
Instead of using `UnresolvedType` as a placeholder for a type hole,
let's switch over to a dedicated "rich" `HoleType` which is capable
of storing "originator" type - type variable or dependent member
type which couldn't be resolved.

This makes it easier for the solver to determine origins of
a hole which helps to diagnose certain problems better. It also
helps code completion to locate "expected type" of the context
even when it couldn't be completely resolved.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Introduce a new type to represent a type hole in constraint system

Instead of using `UnresolvedType` as a placeholder for a type hole,
let's switch over to a dedicated "rich" `HoleType` which is capable
of storing "originator" type - type variable or dependent member
type which couldn't be resolved.

This makes it easier for the solver to determine origins of
a hole which helps to diagnose certain problems better. It also
helps code completion to locate "expected type" of the context
even when it couldn't be completely resolved.

xedin committed:
[AST] NFC: Fix a typo in the `HasTypeHole` description

xedin committed:
[CSApply] Make sure that hole type are not propagated to AST

xedin committed:
[ConstraintSystem] Replace use of `UnresolvedType` with specialized `HoleType`

xedin committed:
[CSDiagnostics] Adjust diagnostics to account that type holes in solution are `UnresolvedType`

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

LucianoPAlmeida commented on include/swift/AST/Types.h:
```diff
@@ -148,7 +148,10 @@ class RecursiveTypeProperties {
     HasOpaqueArchetype   = 0x400,
 
-    Last_Property = HasOpaqueArchetype
+    /// This type constains a type hole.
```
Typo "constains" -> "contains" ? :)

xedin replied:
Thanks!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -2358,6 +2358,13 @@ Type ErrorType::get(Type originalType) {
+Type HoleType::get(ASTContext &ctx, OriginatorType originator) {
+  assert(originator);
+  RecursiveTypeProperties properties = RecursiveTypeProperties::HasTypeHole;
+  auto arena = getArena(properties);
```
I believe this is going to always allocate the type in the permanent arena, which is bad if it contains a pointer to a TypeVariableType. The recursive properties used to determine the arena should be the properties of the originator type, not the new properties of the hole type itself.

xedin replied:
I didn’t realize that, thank you, Slava!

xedin replied:
I need to make sure that types containing holes as well as hole types themselves are allocated in `ConstraintSolver` arena...

xedin replied:
Fixed! Every type which contains holes or is a `HoleType` itself is not going to be allocated in `ConstraintSolver` arena. @slavapestov WDYT?

xedin commented:
@swift-ci please test

xedin committed:
[AST] Make sure that hole types are always allocated in `ConstraintSolver` arena

Since `HoleType` directly as well as other types which could contain holes
are bound to a lifetime of constraint system that created them, we need to
make sure that such types are always allocated using `ConstraintSolver`
arena instead of a permanent one.

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test Linux platform

CodaFi commented:
Outside of the solver, there are a number of places we assert that we aren't seeing types with type variables in them. Those should be extended to also assert about not seeing holes.

xedin commented:
I can do that for completeness and I have added an assert to `SetExprTypes` which makes sure that AST doesn't get holes...

xedin committed:
[CSBindings] Don't attempt to join types that contain holes

theblixguy commented:
Does this allow addressing the following TODO?
https://github.com/apple/swift/blob/83b50cd8e360d78dc320fd70020c5a08899994cb/lib/Sema/CSDiagnostics.cpp#L6328-L6330

xedin commented:
@theblixguy Yes, that's one of the things which could be improved with "rich" hole types.

xedin committed:
[AST/TypeChecker] Add more asserts to make sure types don't get into AST

xedin commented:
@CodaFi Added more asserts to match what we have for type variables.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

davezarzycki commented on lib/AST/Expr.cpp:
```diff
@@ -126,7 +126,7 @@ namespace {
 
 void Expr::setType(Type T) {
-  assert(!T || !T->hasTypeVariable());
+  assert(!T || !T->hasTypeVariable() || !T->hasHole());
```
What do you think about creating a `isCheckedType()` method that ensures that the type is not a `UNCHECKED_TYPE` node? That way we don't need to update every call site that does this: `!T->hasTypeVariable()`.

xedin replied:
That makes sense to me but I think we can't do it at the moment unless `UnresolvedType` is excluded from that check because code completion is still replying on solutions with unresolved types applied to AST (we are currently working on fixing that)...

davezarzycki reviewed (commented):
Sorry for the post-commit review. I didn't see this until now.
